Your suggestions feature + new search UI

miazima opened the pull request.

joelchoo commented on static/templates/your-suggestions-detail.html:
```diff
@@ -0,0 +1,25 @@
+<ion-view class="your-suggestions-detail" view-title="Your Suggestions" hide-back-button="true">
```
Minor nitpick - change your-suggestion**s**-detail to your-suggestion-detail. Same for the filename, and the controller name and where ever it appears

joelchoo commented on beeline/controllers/YourSuggestionsController.js:
```diff
@@ -0,0 +1,36 @@
+  function (
+    // Angular Tools
+    $scope,
+    loadingSpinner,
```
Remove `loadingSpinner` if we're not using it

joelchoo commented on beeline/controllers/YourSuggestionsCreateController.js:
```diff
@@ -0,0 +1,219 @@
+    // Ionic events
+    // ------------------------------------------------------------------------
+    $scope.$on('$ionicView.leave', function () {
+      $scope.resetSuggestion()
```
Document why we are doing this on `ionicView.leave`

joelchoo commented on beeline/controllers/YourSuggestionsCreateController.js:
```diff
@@ -0,0 +1,219 @@
+    // ------------------------------------------------------------------------
+    // Watchers
+    // ------------------------------------------------------------------------
+    $scope.$watch('data.selectedTimeIndex', i => {
```
Should we directly call a function using `ng-change` in `create-new-suggestion.html` rather than changing the value and watching it?

miazima replied:
Yup, calling `ng-change` on user-event would be better practice

miazima commented:
- Renamed `your-suggestions-detail` to `your-suggestion-detail`,  `YourSuggestionsDetailController` to `YourSuggestionDetailController` and `YourSuggestionsCreateController` to `YourSuggestionCreateController`
- Addressed pr comments
- improve styling

joelchoo commented:
@miazima can you list the outstanding features/tasks somewhere? either in this PR or in todoist?

There's definitely the edits in `your-suggestion-detail.html` to deal with the generated route outstanding, but I'm not sure if there's anything else. Thanks!

miazima commented:
TODO (`your-suggestion-detail` page)
1. Retrieve and display generated route
2. Generate user suggested crowdstart route on request
3. Retrieve and display crowdstart route

joelchoo commented on beeline/services/data/SuggestionService.js:
```diff
@@ -0,0 +1,125 @@
+      },
+
+      fetchUserSuggestions: function () {
+        let user = UserService.getUser()
```
more gracefully deal with the case when user is not logged in

joelchoo commented on beeline/controllers/TabsController.js:
```diff
@@ -129,6 +133,14 @@ export default [
     })
 
+    $scope.$watch(() => SuggestionService.getSuggestions(), suggestions => {
+      if (suggestions && suggestions.length > 0) {
```
`$scope.disp.hasSuggestions = (suggestions && suggestions.length > 0)`

joelchoo commented on beeline/controllers/YourSuggestionCreateController.js:
```diff
@@ -0,0 +1,281 @@
+    $scope.resetSuggestion = function () {
+      $scope.data.pickUpLocation = null
+      $scope.data.dropOffLocation = null
+      $scope.data.selectedTimeIndex = 17
```
Add comment on what is special about 17

joelchoo commented on beeline/controllers/YourSuggestionCreateController.js:
```diff
@@ -0,0 +1,281 @@
+        await $ionicPopup.alert({
+          title: 'Error creating suggestion',
+          template: `
+          <div>You have already submitted a similar suggestion!</div>
```
Right now this seems to be checking for exactly the same suggestion. Are we planning to make it check for similar suggestions in the future (i.e. nearby lat lng)? If we are explicitly checking for duplicate suggestions then let's name the function and the error popup accordingly? @chuasweechin your thoughts on the intended UX for this?

joelchoo replied:
Similar seems to suggest it may not be exactly the same

chuasweechin replied:
For this, the intent is to block people from spamming exactly the same suggestion to push up the numbers. I figure that it could be hard to determine false positive using nearby lat long or different timing so we are going for a exact match for now.

chuasweechin replied:
The idea is that we stem the upstream behavior of people spamming suggestion.

joelchoo commented on beeline/controllers/YourSuggestionDetailController.js:
```diff
@@ -0,0 +1,132 @@
+    // force reload when revisit the same route
+    $scope.$on('$ionicView.afterEnter', () => {
+      // $ionicLoading.show({
+      //   template: `<ion-spinner icon='crescent'></ion-spinner><br/><small>Loading route information</small>`,
```
Delete dead code

miazima replied:
This code may need to be revived upon resolving the suggested-routes issue

joelchoo commented on beeline/controllers/YourSuggestionDetailController.js:
```diff
@@ -0,0 +1,132 @@
+          })
+        })
+
+      // $scope.refreshSuggestedRoutes(suggestionId)
```
Dead code

miazima replied:
This fetches the suggested-routes, just have commented out as @chuasweechin wanted to omit that in staging currently

joelchoo commented on beeline/controllers/YourSuggestionsController.js:
```diff
@@ -0,0 +1,36 @@
+    // Angular Tools
+    $scope,
+    SuggestionService,
+    UserService
```
You're not using the `UserService` here. In fact just double check all the controllers to make sure we're only including dependencies that we are using.

joelchoo commented on beeline/directives/locationSearch/locationSearch.js:
```diff
@@ -17,7 +17,7 @@ angular.module('beeline').directive('locationSearch', [
           <div class="input-box" on-tap="openSearch()">
             <div class="input-text placeholder" ng-if="!queryLocation">{{ph}}</div>
-            <div class="input-text" ng-if="queryLocation">{{queryLocation.ADDRESS}}</div>
+            <div class="input-text" ng-if="queryLocation">{{ queryLocation.ADDRESS || queryLocation.BUILDING || queryLocation.ROAD_NAME }}</div>
```
This was my fix for reverse geocode not having the same schema as the search. Just a reminder to myself to check it out.

joelchoo replied:
Checked it out and is fine

joelchoo commented on beeline/directives/locationSelectModal/locationSelectModal.html:
```diff
@@ -36,6 +36,20 @@ <h1 class="title">Search Routes</h1>
+            class="item location-result"
+            on-tap="selectMyLocation()"
+          >
+            {{ data.myLocation.BUILDING || data.myLocation.ROAD_NAME }}
```
Another thing for me to check out about the reverse geocode schema

joelchoo replied:
Checked it out and it's fine

joelchoo commented on beeline/directives/mapCurvedRoute.js:
```diff
@@ -0,0 +1,63 @@
+            opacity: 0.7,
+          }
+
+          // const start = scope.start
```
Dead code

joelchoo commented on beeline/directives/suggestionItem/suggestionItem.html:
```diff
@@ -0,0 +1,36 @@
+    <div class="additional-info">
+      <div class="icon-and-schedule">
+        <span class="icon-span">
+          <!-- <img class="mini-icon" src="img/icon_schedule.svg"> -->
```
Dead code

joelchoo commented on beeline/directives/suggestionItem/suggestionItem.js:
```diff
@@ -0,0 +1,25 @@
+        $scope.parseTime = function (time) {
+          let hours = moment.duration(time).hours()
+          let suffix = hours > 11 ? 'PM' : 'AM'
+          hours = hours > 11 ? hours - 12 : hours
```
Is `00:30` preferred over `12:30`?

miazima replied:
yup should be 12, fixed that

joelchoo commented on beeline/main.js:
```diff
@@ -241,7 +256,7 @@ app
         const [head] = window.document.getElementsByTagName('head')
         head.appendChild(base)
-        $locationProvider.html5Mode({ enabled: true })
+        // $locationProvider.html5Mode({ enabled: true })
```
Still relevant

joelchoo commented on beeline/services/data/SuggestionService.js:
```diff
@@ -0,0 +1,207 @@
+        return createdSuggestion
+      },
+
+      requestCreateNewSuggestion: function (
```
Seems like this should be an "internal" function (not meant to be called directly)? If so, maybe move it out of the return object to become a helper function

joelchoo commented on beeline/services/data/SuggestionService.js:
```diff
@@ -0,0 +1,207 @@
+        return suggestions
+      },
+
+      fetchSuggestions: () => fetchSuggestions(),
```
I think you can just do `fetchSuggestions,` instead?

joelchoo commented on beeline/controllers/YourSuggestionCreateController.js:
```diff
@@ -213,7 +213,7 @@ export default [
       $scope.data.pickUpLocation = null
       $scope.data.dropOffLocation = null
-      $scope.data.selectedTimeIndex = 17
+      $scope.data.selectedTimeIndex = '17' // 8.30 am
```
Wait why is this a string now? Is it meant to be an int or a string?

miazima replied:
It is meant to be a string, similar to how its defined in `$scope.data`, defining it as a integer breaks the behaviour

joelchoo replied:
Wait that's so weird. Why does defining it as an int break stuff?
